src/libcore/ops/drop.rs

```diff
@@ -1,85 +1,120 @@
-/// Used to run some code when a value goes out of scope.
-/// This is sometimes called a 'destructor'.
-///
-/// When a value goes out of scope, it will have its `drop` method called if
-/// its type implements `Drop`. Then, any fields the value contains will also
-/// be dropped recursively.
-///
-/// Because of this recursive dropping, you do not need to implement this trait
-/// unless your type needs its own destructor logic.
-///
-/// Refer to [the chapter on `Drop` in *The Rust Programming Language*][book]
-/// for some more elaboration.
-///
-/// [book]: ../../book/ch15-03-drop.html
-///
-/// # Examples
-///
+/// Destructors for cleaning up resources
+/// 
+/// When a value is no longer needed, Rust will run a "destuctor" on that value.
+/// The most common way that a value is no longer needed is when it goes out of
+/// scope. Destuctors may still run in other circumstances, but we're going to
+/// focus on scope for the examples here. To learn about some of those other cases,
+/// please see [the reference] section on destructors.
+/// 
+/// [the reference]: https://doc.rust-lang.org/reference/destructors.html
+/// 
+/// The `Drop` trait provides a way to implement a custom destructor on a type.
+/// Why would you want such a thing? Well, many types aren't only data: they
+/// manage some sort of resource. That resource may be memory, it may be a file
+/// descriptor, it may be a network socket. But when the type is no longer going
+/// to be used, it should "clean up" that resource by freeing the memory or
+/// closing the file or socket. This is the job of a desctructor, and therefore
+/// the job for `Drop::drop`.
+/// 
 /// ## Implementing `Drop`
-///
-/// The `drop` method is called when `_x` goes out of scope, and therefore
-/// `main` prints `Dropping!`.
-///
-/// ```
+/// 
+/// When a value goes out of scope, it will call `Drop::drop` on that value. For example,
+/// 
+/// ```rust
 /// struct HasDrop;
-///
+/// 
 /// impl Drop for HasDrop {
 ///     fn drop(&mut self) {
 ///         println!("Dropping!");
 ///     }
 /// }
-///
+/// 
 /// fn main() {
 ///     let _x = HasDrop;
 /// }
 /// ```
-///
-/// ## Dropping is done recursively
-///
-/// When `outer` goes out of scope, the `drop` method will be called first for
-/// `Outer`, then for `Inner`. Therefore, `main` prints `Dropping Outer!` and
-/// then `Dropping Inner!`.
-///
-/// ```
-/// struct Inner;
-/// struct Outer(Inner);
-///
-/// impl Drop for Inner {
-///     fn drop(&mut self) {
-///         println!("Dropping Inner!");
-///     }
+/// 
+/// Here, `main` will print `Dropping!`. In other words, the compiler generates code that
+/// kind of looks like this:
+/// 
+/// ```rust,ignore
+/// fn main() {
+///     let mut _x = HasDrop;
+/// 
+///     Drop::drop(&mut x);
```

Comment:
Programmer Analogy Pedantic Warning

I worry that this example might lead people to think "oh, it's mutable the whole time?" An alternative would be

```rust
fn main() {
    let _x = HasDrop;

    let mut _x = _x;
    Drop::drop(&mut x);
}
```

This is also not strictly valid because the value isn't moved.

Reply:
i like it